src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -972,15 +972,21 @@ private static int CanonicallyCompareDiagnostics(Diagnostic x, Diagnostic y)
 
             int codeCompare = xCode.CompareTo(yCode);
 
-            // ToString fails for a diagnostic with an error code that does not prevent successful delegate conversion.
-            // Also, the order doesn't matter, since all such diagnostics will be dropped.
-            if (!ErrorFacts.PreventsSuccessfulDelegateConversion(xCode) || !ErrorFacts.PreventsSuccessfulDelegateConversion(yCode))
-            {
+            // Optimization: don't bother 
+            if (codeCompare != 0)
                 return codeCompare;
+
+            for (int i = 0; i < Math.Min(x.Arguments.Count, y.Arguments.Count); i++)
+            {
+                object argx = x.Arguments[i];
+                object argy = y.Arguments[i];
+
+                codeCompare = string.CompareOrdinal(argx?.ToString(), argy?.ToString());
```

Comment:
just to check but string.CompareOrdinal works on null values?

Reply:
Yep